overlord/exportstate/special.go

```diff
@@ -0,0 +1,208 @@
+		ConsumerIsHost: false, // Those tools are for other snaps to use.
+		Exports:        make(map[string]ExportedFile, len(tools)),
+	}
+	for _, tool := range tools {
+		set.Exports[tool] = ExportedFile{
+			Name:       tool,
+			SourcePath: filepath.Join(basePath, tool),
+		}
+	}
+	return set
+}
+
+// manifestForCoreSystem returns the manifest of the host as seen on ubuntu-core systems.
+func manifestForCoreSystem() *Manifest {
+	return &Manifest{
+		ExportedName:    "snapd",
+		ExportedVersion: "host",
+		// There are no export sets here, snapd snap is going to provide the tools.
+	}
+}
+
+// manifestForClassicSystem returns the manifest of the host as seen on classic systems.
+func manifestForClassicSystem() *Manifest {
+	tools := snapdTools(dirs.DistroLibExecDir)
+	return &Manifest{
+		ExportedName:    "snapd",
+		ExportedVersion: "host",
+		Sets:            map[string]ExportSet{tools.Name: tools},
+	}
+}
+
+// manifestForSnapdSnap returns the manifest of the snapd snap.
+func manifestForSnapdSnap(info *snap.Info) *Manifest {
+	exportedName, exportedVersion := exportedNameVersionForSnapd(info)
+	tools := snapdTools("/usr/lib/snapd")
+	return &Manifest{
+		ExportedName:     exportedName,
+		ExportedVersion:  exportedVersion,
+		SnapInstanceName: info.InstanceName(),
+		SnapRevision:     info.Revision.String(),
+		Sets:             map[string]ExportSet{tools.Name: tools},
+	}
+}
+
+// manifestForCoreSnap returns the manifest of the core snap.
+func manifestForCoreSnap(info *snap.Info) *Manifest {
+	exportedName, exportedVersion := exportedNameVersionForCore(info)
+	tools := snapdTools("/usr/lib/snapd")
+	return &Manifest{
+		ExportedName:     exportedName,
+		ExportedVersion:  exportedVersion,
+		SnapInstanceName: info.InstanceName(),
+		SnapRevision:     info.Revision.String(),
+		Sets:             map[string]ExportSet{tools.Name: tools},
+	}
+}
+
+// manifestForRegularSnap returns the manifest for a snap other than core or snapd.
+func manifestForRegularSnap(info *snap.Info) *Manifest {
+	exportedName, exportedVersion := exportedNameVersionForRegularSnap(info)
+	return &Manifest{
+		ExportedName:     exportedName,
+		ExportedVersion:  exportedVersion,
+		SnapInstanceName: info.InstanceName(),
+		SnapRevision:     info.Revision.String(),
+		// TODO: eventually get this from the snap.yaml
+	}
+}
+
+func exportedNameVersionForSnapd(info *snap.Info) (exportedName string, exportedVersion string) {
+	return "snapd", info.Revision.String()
+}
+
+func exportedNameVersionForCore(info *snap.Info) (exportedName string, exportedVersion string) {
+	return "snapd", fmt.Sprintf("core_%s", info.Revision)
+}
+
+func exportedNameVersionForRegularSnap(info *snap.Info) (exportedName string, exportedVersion string) {
+	if info.SnapName() == "core" || info.SnapName() == "snapd" {
+		panic("internal error, cannot use exportedNameVersionForRegularSnap with core or snapd")
+	}
+	exportedName = info.SnapName() // Instance key goes to exportedVersion
```

Comment:
because we have just one current at the level of /export/snap  I don't think putting instances all under that level makes sense. We really need to structure the tree like:

../export/instanceName/revision

then snapd has some special casing for host and core but it alone.

Reply:
I haven't touched this yet. I think we can discuss if this is worth doing now. Technically we can leave it or change it and there's no difference because we are not allowing instances on the snaps with exported content.

We should still think about this today and pick the direction. It should be easy to change either way.

Reply:
This is now done.